plugins/inputs/gnmi/README.md

```diff
@@ -32,6 +32,9 @@ It has been optimized to support gNMI telemetry as produced by Cisco IOS XR
   ## enable client-side TLS and define CA to authenticate the device
   # enable_tls = true
   # tls_ca = "/etc/telegraf/ca.pem"
+  ## Minimal TLS version to accept by the client
+  ## If undet, the value defaults to Golang's default which is TLS v1.2 currently.
```

Comment:
I don't think we should say in each config file what the minimum TLS version is. It is asking for trouble if/when it changes and we do not catch that.

under -> unset

Reply:
I can see the argument on either side for putting this in each plugin that uses it, however it feels like putting it only in the global `[agent]` section might make the most sense. Eg:
```
[agent]
...
## Minimal TLS version to use for plugins that support TLS.
## If unset, the value defaults to Golang's default which is TLS v1.2 for clients and v1.0 for servers.
## Telegraf typically acts as a client but will act as a server for listening plugins that use HTTP.
# tls_min_version = ""
```

Like with other directives, I can see it being useful for individual plugins to be able to override the agent default. This will allow people to say, set the default to 1.2 and then override in an individual plugin to use 1.1 for some old server they are in the process of upgrading.

Reply:
I can see arguments on both sides, but what if we adjusted the global `[agent]` configuration and have it apply to all plugins?
```
[agent]
...
## Minimal TLS version to use for plugins that support TLS.
## If unset, the value defaults to Golang's default which is TLS v1.2 for clients and v1.0
## for servers. Telegraf typically acts as a client but will act as a server for listening
## plugins that use HTTP.
# tls_min_version = ""
```

Like other directives, users could then set `tls_min_version` in individual plugins which would allow using 1.2 everywhere except, for example, that one plugin that is talking to a 1.1 server (so the user gains the benefit of min 1.2 while in the process of upgrading that 1.1 server).

Reply:
I thought about having a global option, but that has some issues as the minimum TLS version can also be set for servers (e.g for `http_listener_v2`) and this causes some ambiguity.

> Like other directives, users could then set tls_min_version in individual plugins which would allow using 1.2 everywhere except, for example, that one plugin that is talking to a 1.1 server (so the user gains the benefit of min 1.2 while in the process of upgrading that 1.1 server).

Well that is the case currently. Golang defaults to TLS v1.2 if you do not state otherwise, so we only deviate from v1.2 if explicitly set by the user.

Reply:
> I thought about having a global option, but that has some issues as the minimum TLS version can also be set for servers (e.g for `http_listener_v2`) and this causes some ambiguity.

Where is this limitation? It shouldn't be a Golang limitation as https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/crypto/tls/common.go;l=685 says:

```
	// MinVersion contains the minimum TLS version that is acceptable.
	//
	// By default, TLS 1.2 is currently used as the minimum when acting as a
	// client, and TLS 1.0 when acting as a server. TLS 1.0 is the minimum
	// supported by this package, both as a client and as a server.
	//
	// The client-side default can temporarily be reverted to TLS 1.0 by
	// including the value "x509sha1=1" in the GODEBUG environment variable.
	// Note that this option will be removed in Go 1.19 (but it will still be
	// possible to set this field to VersionTLS10 explicitly).
```

This means that we should be able to set clients to TLS v1.1 and higher. I agree that trying to also set GODEBUG to enable v1.0 in clients is pointless since that is going away in 1.19 (I missed this 1.19 detail before). Due note that the option is also useful for hardening and requiring v1.3 ahead of Golang defaults (this is considered industry best practice but I agree is not the correct default for telegraf at this time). 

> > Like other directives, users could then set tls_min_version in individual plugins which would allow using 1.2 everywhere except, for example, that one plugin that is talking to a 1.1 server (so the user gains the benefit of min 1.2 while in the process of upgrading that 1.1 server).
> 
> Well that is the case currently. Golang defaults to TLS v1.2 if you do not state otherwise, so we only deviate from v1.2 if explicitly set by the user.

As mentioned elsewhere, v1.2 is only the Golang default when acting as a client and as you pointed out, `http_listener_v2` is when acting as a server. As such, we still have a low security issue for when telegraf is acting as a server and using a too low minimum version.

My suggestion is about being clear and explicitly setting v1.2 in the global section which:
* addresses the too-low Golang default of v1.0 for servers by setting it to v1.2
* makes it discoverable to users what telegraf is doing (ie, they don't have to know that Golang 1.18 changed the client default and that telegraf is using 1.18; they just see it in the config)
* conveniently allows users to adjust the default for all plugins to use v1.3 to harden their configuration and meet certain certification/operational requirements

As for ambiguity of server vs client telegraf default in the global section, I think we can make things clear for users with appropriate documentation. My previous suggestion as not is good as it could be. Perhaps:

```
[agent]
...
## Minimal TLS version to use for plugins that support TLS. Please
## choose one of: TLS13, TLS12 or TLS11. TLS12 is recommended for widest
## secure compatibility. Note that telegraf typically acts as a client
## but will act as a server for listening plugins that use HTTP. TLS10 is
## only supported when acting as a server, so if TLS10 is needed, please
## set in the individual plugin.
## 
tls_min_version = "TLS12"
```

(this implies that only TLS11 and higher is supported in the global section with support for setting TLS10 in the listeners' configuration. If you'd prefer to allow setting TLS10 in the global section, the comment should be adjusted that clients will be upgraded to TLS11).

Reply:
My point is not on limitations it's about the ambiguity of having __one__ `tls_min_version`. As you suggest you can make it apply to both client and server side, but that would break backward compatibility for all server sides at least.

So my fear is the following scenario: People have some clients and some servers in their config and for whatever reason are constraint to < v1.2 on the server side. Instead of changing the setting in individual plugins, they will much likely change the global option to say v1.0 and be more insecure than before... I'd rather prefer to set v1.2 as a default _without_ an option, _forcing_ people to change it in the individual plugins and use the innate laziness.. ;-)

Reply:
Pushed for making v1.2 the (explicit) default for both server and client. Allowing to revert on a plugin basis to mitigate the risk of users lowering the global options out of laziness...